exec/cnex/lib/binary.c

```diff
@@ -0,0 +1,478 @@
+#include "binary.h"
+
+#include <assert.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cell.h"
+#include "exec.h"
+#include "number.h"
+#include "stack.h"
+#include "nstring.h"
+
+
+static void range_check32(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_INT32) || number_is_greater(x, BID_MAX_INT32)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+static void range_checkU32(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_UINT32) || number_is_greater(x, BID_MAX_UINT32)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+static void range_check64(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_INT64) || number_is_greater(x, BID_MAX_INT64)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+static void range_checkU64(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_UINT64) || number_is_greater(x, BID_MAX_UINT64)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+void binary_and32(TExecutor *exec)
+{
+    Number y = top(exec->stack)->number; pop(exec->stack);
+    Number x = top(exec->stack)->number; pop(exec->stack);
+
+    range_check32(exec, x);
```

Comment:
If `range_check32()` raises an exception, you need to stop what you're doing and exit this function. You can't just proceed and push the result on the stack, because the stack has already been adjusted to match what the exception handler expects (and the top of the stack is supposed to be the exceptioninfo record).

Reply:
Fixed.  range_check()'s now return a BOOL, and exit the module function if an exception has been raised.